subprojects/core-api/src/main/java/org/gradle/api/artifacts/ComponentMetadataDetails.java

```diff
@@ -105,6 +106,20 @@
      */
     void addVariant(String name, String base, Action<? super VariantMetadata> action);
 
+    /**
+     * This is the lenient version of {@link #addVariant(String, String, Action)}.
+     * The only difference is that this will do nothing (instead of throwing an error), if the 'base' variant does not exist for a component.
+     * This is particularly useful for rules that are applied to {@link org.gradle.api.artifacts.dsl.ComponentMetadataHandler#all(Class)} components.
+     *
+     * @param name a name for the variant
+     * @param base name of the variant (pom or Gradle module metadata) or configuration (ivy.xml metadata) from which the new variant will be initialized
+     * @param action the action to populate the variant
+     *
+     * @since 6.1
+     */
+    @Incubating
+    void maybeAddVariant(String name, String base, Action<? super VariantMetadata> action);
```

Comment:
:o: Consider returning a `boolean` here so that users can know if the variant was created (`true`) or not (`false`)

Reply:
While I think in theory we could make it work and it could have a certain use, it would be a quite involved change. If we want to know here, if the addition is successful or not, we need to make suer that the method (which so far only registers the rule but does not execute the addition) runs **after** the variant derivation for pom has been performed. Currently this is all done and chained together lazily later if needed.
Apart from that I don't know how keen I am on enabling the use of if/else constructs here. I would rather evolve the API in the future to be even more _matching-oriented_ e.g. by matching on attributes instead of name.

Reply:
Ah right, forgot about that aspect.
Then leave it as-is